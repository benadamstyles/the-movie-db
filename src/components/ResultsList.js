import React, {Component} from 'react'
import styled from 'react-emotion'
import Async from 'react-async'
import PropTypes from 'prop-types'
import ResultsListItem, {Placeholder} from './ResultsListItem'
import {fetch} from '../util/fetch'
import OnScrollEnd from './OnScrollEnd'

const List = styled('ul')`
  list-style-type: none;
  padding: 0;
  margin: 0;
`

const dataHasResults = data => data && data.results && data.results.length > 0

export default class ResultsList extends Component {
  static propTypes = {
    searchText: PropTypes.string.isRequired,
    onSelectMovie: PropTypes.func.isRequired,
  }

  shouldComponentUpdate(nextProps) {
    return this.props.searchText !== nextProps.searchText
  }

  fetchResults = page =>
    this.props.searchText
      ? fetch(
          `https://api.themoviedb.org/3/search/movie?api_key=ecfdccbb9349ca6c7898c6ec1b482560&query=${
            this.props.searchText
          }&page=${page}`
        )
          .then(res => res.json())
          .then(res => ({
            page: res.page,
            results: res.results,
            total_pages: res.total_pages,
          }))
      : Promise.resolve([])

  nextPage = async prevData => {
    const nextPage = prevData.page + 1
    if (nextPage <= prevData.total_pages) {
      const nextData = await this.fetchResults(nextPage)
      return {
        page: nextPage,
        total_pages: prevData.total_pages,
        results: [
          ...prevData.results,

          // NOTE: this fixes a movieDB bug where duplicates are returned
          ...nextData.results.filter(
            ({id}) => !prevData.results.some(prevMovie => prevMovie.id === id)
          ),
        ],
      }
    } else {
      return prevData
    }
  }

  initialFetch = () => this.fetchResults(1)

  render() {
    return (
      <Async
        promiseFn={this.initialFetch}
        deferFn={this.nextPage}
        watch={this.props.searchText}>
        {({isLoading, run, data}) => (
          <>
            <OnScrollEnd callback={() => run(data)} />

            <Async.Loading>
              <Placeholder>loading ...</Placeholder>
            </Async.Loading>

            <List>
              {dataHasResults(data) &&
                data.results.map(movie => (
                  <ResultsListItem
                    key={String(movie.id)}
                    movie={movie}
                    onSelectMovie={this.props.onSelectMovie}
                  />
                ))}

              <Async.Resolved>
                {this.props.searchText &&
                  !dataHasResults(data) && (
                    <Placeholder>[No results found]</Placeholder>
                  )}
              </Async.Resolved>
            </List>
          </>
        )}
      </Async>
    )
  }
}
