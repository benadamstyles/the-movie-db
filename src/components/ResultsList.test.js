import React from 'react'
import {render, fireEvent, cleanup, waitForElement} from 'react-testing-library'
import ResultsList from './ResultsList'

jest.mock('../util/fetch', () => {
  const mockMovie = {
    poster_path: '/cezWGskPY5x7GaglTTRN4Fugfb8.jpg',
    adult: false,
    overview: 'This is a mock movie.',
    release_date: '2012-04-25',
    genre_ids: [878, 28, 12],
    id: 0,
    original_title: 'Mock Movie',
    original_language: 'en',
    title: 'Mock Movie',
    backdrop_path: '/hbn46fQaRmlpBuUrEiFqv0GDL6Y.jpg',
    popularity: 7.353212,
    vote_count: 8503,
    video: false,
    vote_average: 7.33,
  }

  return {
    fetch: () =>
      Promise.resolve({
        json: () =>
          Promise.resolve({
            results: [...Array(5).keys()].map((_, id) => ({
              ...mockMovie,
              id,
            })),
          }),
      }),
  }
})

afterEach(cleanup)

describe('ResultsList', () => {
  it('renders a loading message while fetching', () => {
    const {container, getByText} = render(
      <ResultsList searchText="abc" onSelectMovie={jest.fn()} />
    )
    expect(container).toMatchSnapshot()
    expect(getByText('loading ...')).toBeTruthy()
  })

  it('renders a list of movies after a successful fetch', async () => {
    const {getAllByTestId, container} = render(
      <ResultsList searchText="abc" onSelectMovie={jest.fn()} />
    )
    await new Promise(r => setTimeout(r, 10))
    expect(container).toMatchSnapshot()
    expect(getAllByTestId('result-title')).toHaveLength(5)
  })
})
