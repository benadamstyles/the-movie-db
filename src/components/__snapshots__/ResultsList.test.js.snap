// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsList renders a list of movies after a successful fetch 1`] = `
<div>
  <ul
    class="css-ewdv3l"
  >
    <li>
      <button
        class="css-rg16dp"
        type="button"
      >
        <span
          class="css-17vo2iu"
          data-testid="result-title"
        >
          Mock Movie
        </span>
      </button>
    </li>
    <li>
      <button
        class="css-rg16dp"
        type="button"
      >
        <span
          class="css-17vo2iu"
          data-testid="result-title"
        >
          Mock Movie
        </span>
      </button>
    </li>
    <li>
      <button
        class="css-rg16dp"
        type="button"
      >
        <span
          class="css-17vo2iu"
          data-testid="result-title"
        >
          Mock Movie
        </span>
      </button>
    </li>
    <li>
      <button
        class="css-rg16dp"
        type="button"
      >
        <span
          class="css-17vo2iu"
          data-testid="result-title"
        >
          Mock Movie
        </span>
      </button>
    </li>
    <li>
      <button
        class="css-rg16dp"
        type="button"
      >
        <span
          class="css-17vo2iu"
          data-testid="result-title"
        >
          Mock Movie
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`ResultsList renders a loading message while fetching 1`] = `
<div>
  <p
    class="css-7n945j"
  >
    loading ...
  </p>
  <ul
    class="css-ewdv3l"
  />
</div>
`;
